File: backend-node/src/routes/costCenters.ts
import { Router } from 'express'
import { authMiddleware, requireRole } from '../middleware/auth.ts'
import { query, queryOne } from '../db.ts'
import { logAction } from '../services/audit.ts'
import type { AuthRequest } from '../types.ts'

const router = Router()

// Build full cost center response including members
async function getCostCenter(id: string) {
  const cc = await queryOne(
    'SELECT id, code, name, is_active, created_at FROM cost_centers WHERE id = $1',
    [id]
  )
  if (!cc) return null
  const members = await query(
    `SELECT ccm.id, ccm.role,
            u.id AS user_id, u.email, u.display_name, u.role AS user_role, u.is_active
     FROM cost_center_members ccm
     JOIN users u ON u.id = ccm.user_id
     WHERE ccm.cost_center_id = $1`,
    [id]
  )
  return {
    ...cc,
    members: members.map((m) => ({
      id: m.id,
      role: m.role,
      user: { id: m.user_id, email: m.email, display_name: m.display_name, role: m.user_role, is_active: m.is_active },
    })),
  }
}

router.get('/cost-centers', authMiddleware, async (req, res) => {
  const user = (req as AuthRequest).user
  let rows: Array<Record<string, unknown>>

  if (user.role === 'admin' || user.role === 'finance') {
    rows = await query('SELECT id FROM cost_centers WHERE is_active = true ORDER BY code')
  } else {
    rows = await query(
      `SELECT cc.id FROM cost_centers cc
       JOIN cost_center_members ccm ON ccm.cost_center_id = cc.id
       WHERE ccm.user_id = $1 AND cc.is_active = true ORDER BY cc.code`,
      [user.id]
    )
  }

  const costCenters = await Promise.all(rows.map((r) => getCostCenter(r.id as string)))
  res.json(costCenters.filter(Boolean))
})

router.post('/cost-centers', authMiddleware, requireRole('admin'), async (req, res) => {
  const { code, name } = req.body as { code: string; name: string }
  if (!code || !name) { res.status(400).json({ detail: 'code and name required' }); return }

  const existing = await queryOne('SELECT id FROM cost_centers WHERE code = $1', [code])
  if (existing) { res.status(409).json({ detail: 'Cost center code already exists' }); return }

  const rows = await query(
    'INSERT INTO cost_centers (code, name) VALUES ($1, $2) RETURNING id',
    [code, name]
  )
  const user = (req as AuthRequest).user
  await logAction({ entityType: 'CostCenter', entityId: rows[0].id as string, action: 'created', userId: user.id, details: { code, name } })
  res.status(201).json(await getCostCenter(rows[0].id as string))
})

router.put('/cost-centers/:id', authMiddleware, requireRole('admin'), async (req, res) => {
  const { name, is_active } = req.body as { name?: string; is_active?: boolean }
  const sets: string[] = ['updated_at = NOW()']
  const vals: unknown[] = []
  if (name !== undefined) { vals.push(name); sets.push(`name = $${vals.length}`) }
  if (is_active !== undefined) { vals.push(is_active); sets.push(`is_active = $${vals.length}`) }
  vals.push(req.params.id)
  await query(`UPDATE cost_centers SET ${sets.join(', ')} WHERE id = $${vals.length}`, vals)
  const cc = await getCostCenter(req.params.id)
  if (!cc) { res.status(404).json({ detail: 'Not found' }); return }
  res.json(cc)
})

router.post('/cost-centers/:id/members', authMiddleware, requireRole('admin'), async (req, res) => {
  const { user_id, role = 'owner' } = req.body as { user_id: string; role?: string }
  const ccId = req.params.id
  const actor = (req as AuthRequest).user

  const ccExists = await queryOne('SELECT id FROM cost_centers WHERE id = $1', [ccId])
  if (!ccExists) { res.status(404).json({ detail: 'Cost center not found' }); return }
  const userExists = await queryOne('SELECT id FROM users WHERE id = $1', [user_id])
  if (!userExists) { res.status(404).json({ detail: 'User not found' }); return }

  await query(
    `INSERT INTO cost_center_members (cost_center_id, user_id, role)
     VALUES ($1, $2, $3)
     ON CONFLICT (cost_center_id, user_id) DO UPDATE SET role = $3, updated_at = NOW()`,
    [ccId, user_id, role]
  )
  await logAction({ entityType: 'CostCenter', entityId: ccId, action: 'member_added', userId: actor.id, details: { user_id, role } })
  res.status(201).json(await getCostCenter(ccId))
})

router.delete('/cost-centers/:id/members/:userId', authMiddleware, requireRole('admin'), async (req, res) => {
  const { id: ccId, userId } = req.params
  const actor = (req as AuthRequest).user
  const result = await query(
    'DELETE FROM cost_center_members WHERE cost_center_id = $1 AND user_id = $2 RETURNING id',
    [ccId, userId]
  )
  if (!result.length) { res.status(404).json({ detail: 'Member not found' }); return }
  await logAction({ entityType: 'CostCenter', entityId: ccId, action: 'member_removed', userId: actor.id, details: { removed_user_id: userId } })
  res.status(204).send()
})

export default router

File: backend/app/routers/cost_centers.py
from uuid import UUID

from fastapi import APIRouter, Depends, HTTPException
from sqlalchemy.orm import Session

from ..database import get_db
from ..dependencies import get_current_user, require_admin
from ..models.cost_center import CostCenter, CostCenterMember, MemberRole
from ..models.user import User, UserRole
from ..schemas.cost_center import CostCenterCreate, CostCenterRead, CostCenterUpdate, MemberAdd
from ..services.audit import log_action

router = APIRouter(prefix="/cost-centers", tags=["cost-centers"])


def _assert_access(cost_center: CostCenter, user: User) -> None:
    if user.role in (UserRole.admin, UserRole.finance):
        return
    ids = [m.user_id for m in cost_center.members]
    if user.id not in ids:
        raise HTTPException(status_code=403, detail="Not a member of this cost center")


@router.get("", response_model=list[CostCenterRead])
def list_cost_centers(
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    if current_user.role in (UserRole.admin, UserRole.finance):
        return db.query(CostCenter).filter(CostCenter.is_active == True).all()

    memberships = current_user.cost_center_memberships
    cc_ids = [m.cost_center_id for m in memberships]
    return db.query(CostCenter).filter(CostCenter.id.in_(cc_ids), CostCenter.is_active == True).all()


@router.post("", response_model=CostCenterRead, status_code=201)
def create_cost_center(
    body: CostCenterCreate,
    db: Session = Depends(get_db),
    current_user: User = Depends(require_admin),
):
    if db.query(CostCenter).filter(CostCenter.code == body.code).first():
        raise HTTPException(status_code=409, detail="Cost center code already exists")
    cc = CostCenter(code=body.code, name=body.name)
    db.add(cc)
    db.commit()
    db.refresh(cc)
    log_action(db, entity_type="CostCenter", entity_id=cc.id, action="created",
               user_id=current_user.id, details={"code": cc.code, "name": cc.name})
    db.commit()
    return cc


@router.put("/{cc_id}", response_model=CostCenterRead)
def update_cost_center(
    cc_id: UUID,
    body: CostCenterUpdate,
    db: Session = Depends(get_db),
    current_user: User = Depends(require_admin),
):
    cc = db.get(CostCenter, cc_id)
    if not cc:
        raise HTTPException(status_code=404, detail="Cost center not found")
    if body.name is not None:
        cc.name = body.name
    if body.is_active is not None:
        cc.is_active = body.is_active
    db.commit()
    db.refresh(cc)
    return cc


@router.post("/{cc_id}/members", response_model=CostCenterRead, status_code=201)
def add_member(
    cc_id: UUID,
    body: MemberAdd,
    db: Session = Depends(get_db),
    current_user: User = Depends(require_admin),
):
    cc = db.get(CostCenter, cc_id)
    if not cc:
        raise HTTPException(status_code=404, detail="Cost center not found")
    user = db.get(User, body.user_id)
    if not user:
        raise HTTPException(status_code=404, detail="User not found")

    existing = (
        db.query(CostCenterMember)
        .filter(CostCenterMember.cost_center_id == cc_id, CostCenterMember.user_id == body.user_id)
        .first()
    )
    if existing:
        existing.role = body.role
    else:
        db.add(CostCenterMember(cost_center_id=cc_id, user_id=body.user_id, role=body.role))

    log_action(db, entity_type="CostCenter", entity_id=cc_id, action="member_added",
               user_id=current_user.id, details={"member_user_id": str(body.user_id), "role": body.role})
    db.commit()
    db.refresh(cc)
    return cc


@router.delete("/{cc_id}/members/{user_id}", status_code=204)
def remove_member(
    cc_id: UUID,
    user_id: UUID,
    db: Session = Depends(get_db),
    current_user: User = Depends(require_admin),
):
    member = (
        db.query(CostCenterMember)
        .filter(CostCenterMember.cost_center_id == cc_id, CostCenterMember.user_id == user_id)
        .first()
    )
    if not member:
        raise HTTPException(status_code=404, detail="Member not found")
    db.delete(member)
    log_action(db, entity_type="CostCenter", entity_id=cc_id, action="member_removed",
               user_id=current_user.id, details={"removed_user_id": str(user_id)})
    db.commit()
